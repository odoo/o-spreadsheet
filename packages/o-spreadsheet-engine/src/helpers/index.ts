export * from "./misc";
export * from "./numbers";
export * from "./references";
export * from "./uuid";
