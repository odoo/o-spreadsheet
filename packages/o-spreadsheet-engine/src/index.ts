export { FunctionCodeBuilder, Scope, type FunctionCode } from "./formulas/code_builder";
export {
  OPERATOR_MAP,
  UNARY_OPERATOR_MAP,
  compile,
  compileTokens,
  functionCache,
} from "./formulas/compiler";
export {
  OP_PRIORITY,
  convertAstNodes,
  iterateAstNodes,
  mapAst,
  parse,
  parseTokens,
} from "./formulas/parser";
export type {
  AST,
  ASTFuncall,
  ASTOperation,
  ASTString,
  ASTSymbol,
  ASTUnaryOperation,
} from "./formulas/parser";
export { rangeTokenize } from "./formulas/range_tokenizer";
export { POSTFIX_UNARY_OPERATORS, tokenize } from "./formulas/tokenizer";
export type { Token } from "./formulas/tokenizer";
export { FunctionRegistry } from "./functions/function_registry";
export { categories } from "./functions/functionRegistry";
export * from "./helpers";
export { Model } from "./model";
export { BasePlugin } from "./plugins/base_plugin";
export { Registry } from "./registry";
export { StateObserver, type StateObserverChange } from "./state_observer";
export * from "./translation";
export { CellValue } from "./types/cells";
export { SpreadsheetClipboardData } from "./types/clipboard";
export { CoreGetters, PluginGetters } from "./types/coreGetters";
export * from "./types/errors";
export { Format } from "./types/format";
export * from "./types/history";
export * from "./types/locale";
export * from "./types/misc";
export * from "./types/validator";
export const __info__ = {};
