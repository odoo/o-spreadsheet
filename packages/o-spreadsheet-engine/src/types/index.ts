export * from "./base";
export * from "./errors";
export * from "./functions";
export * from "./history";
export * from "./isMatrix";
export * from "./locale";
export * from "./validator";
