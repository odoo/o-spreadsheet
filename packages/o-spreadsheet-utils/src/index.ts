export * from "./translation";
export * from "./uuid";
