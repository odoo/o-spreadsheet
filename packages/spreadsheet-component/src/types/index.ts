/**
 * State
 *
 * This file defines the basic types involved in maintaining the running state
 * of a o-spreadsheet.
 *
 * The most important exported values are:
 * - interface GridState: the internal type of the state managed by the model
 */

import { Chart } from "chart.js";

export * from "./autofill";
export * from "./cells";
export * from "./chart/chart";
export * from "./clipboard";
export * from "./collaborative/revisions";
export * from "./collaborative/session";
export * from "./commands";
export * from "./conditional_formatting";
export * from "./currency";
export * from "./data_validation";
export * from "./env";
export * from "./figure";
export * from "./format";
export * from "./functions";
export * from "./getters";
export * from "./history";
export * from "./locale";
export * from "./misc";
export * from "./pivot";
export * from "./pivot_runtime";
export * from "./range";
export * from "./rendering";
export * from "./table";
export * from "./workbook_data";

declare global {
  interface Window {
    Chart: typeof Chart;
  }
}
