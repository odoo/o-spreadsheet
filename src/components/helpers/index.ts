export * from "@odoo/o-spreadsheet-engine/components/helpers/css";
