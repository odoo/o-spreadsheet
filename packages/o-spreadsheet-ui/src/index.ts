import * as Engine from "@odoo/o-spreadsheet-engine";

export function salut() {
  console.log("Salut !");
}

export function createSpreadsheetModel() {
  return Engine.SpreadsheetModel();
}