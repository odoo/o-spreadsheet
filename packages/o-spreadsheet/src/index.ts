import * as UI from "@odoo/o-spreadsheet-ui";


export { UI };