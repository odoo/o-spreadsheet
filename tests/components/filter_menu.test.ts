import { Model } from "../../src";
import { UID } from "../../src/types";
import {
  createFilter,
  hideRows,
  setCellContent,
  setFormat,
  updateFilter,
} from "../test_helpers/commands_helpers";
import { keyDown, simulateClick } from "../test_helpers/dom_helper";
import { getCellsObject, mountSpreadsheet, nextTick, target } from "../test_helpers/helpers";

async function openFilterMenu() {
  await simulateClick(".o-filter-icon");
}

describe("Filter menu component", () => {
  let fixture: HTMLElement;
  let model: Model;
  let sheetId: UID;

  function getFilterMenuValues() {
    const values: { value: string; isChecked: boolean }[] = [];
    const filterValueEls = fixture.querySelectorAll(".o-filter-menu-value");
    for (const filterValue of filterValueEls) {
      const isChecked = !!filterValue.querySelector(".o-filter-menu-value-checked span")
        ?.textContent;
      const value = filterValue.querySelector(".o-filter-menu-value-text")!.textContent!;
      values.push({ value, isChecked });
    }
    return values;
  }

  beforeEach(async () => {
    ({ model, fixture } = await mountSpreadsheet());
    sheetId = model.getters.getActiveSheetId();
  });

  describe("Filter Tests", () => {
    beforeEach(async () => {
      createFilter(model, "A1:A5");
      setCellContent(model, "A1", "header");
      setCellContent(model, "A2", "1");
      setCellContent(model, "A3", "1");
      setCellContent(model, "A4", "2");

      createFilter(model, "B1:B4");
      setCellContent(model, "B2", "B2");
      setCellContent(model, "B3", "B3");
      setCellContent(model, "B4", "B4");
      await nextTick();
    });

    test("Filter menu is correctly rendered", async () => {
      await openFilterMenu();
      expect(fixture.querySelector(".o-filter-menu")).toMatchSnapshot();
    });

    test("Duplicates values are not displayed twice", async () => {
      await openFilterMenu();
      const values = getFilterMenuValues();
      expect(values.map((val) => val.value)).toEqual(["(Blanks)", "1", "2"]);
    });

    test("Opening the filter menu of another filter update the values", async () => {
      const filterIcons = fixture.querySelectorAll(".o-filter-icon");
      await simulateClick(filterIcons[0]);
      let values = getFilterMenuValues();
      expect(values.map((val) => val.value)).toEqual(["(Blanks)", "1", "2"]);

      await simulateClick(filterIcons[1]);
      values = getFilterMenuValues();
      expect(values.map((val) => val.value)).toEqual(["B2", "B3", "B4"]);
    });

    test("Uppercase/Lowercase are not displayed twice. Only the first occurrence is displayed", async () => {
      setCellContent(model, "A2", "PercevaL");
      setCellContent(model, "A3", "perceval");
      setCellContent(model, "A4", "gauviN");
      setCellContent(model, "A5", "Gauvin");

      await openFilterMenu();
      const values = getFilterMenuValues();
      expect(values.map((val) => val.value)).toEqual(["gauviN", "PercevaL"]);
    });

    test("Displayed values are sorted using natural sorting", async () => {
      // Sort order should be 2,10,a,B and not 10,2,B,a (the default string comparison)
      setCellContent(model, "A2", "10");
      setCellContent(model, "A3", "2");
      setCellContent(model, "A4", "B");
      setCellContent(model, "A5", "a");

      await openFilterMenu();
      const values = getFilterMenuValues();
      expect(values.map((val) => val.value)).toEqual(["2", "10", "a", "B"]);
    });

    test("We display the formated value of the cells", async () => {
      setFormat(model, "m/d/yyyy", target("A4"));
      await openFilterMenu();
      const values = getFilterMenuValues();
      expect(values.map((val) => val.value)).toEqual(["(Blanks)", "1", "1/1/1900"]);
    });

    test("Values are checked depending on the filter state", async () => {
      updateFilter(model, "A1", ["1"]);
      await openFilterMenu();
      const values = getFilterMenuValues();
      expect(values).toEqual([
        { value: "(Blanks)", isChecked: true },
        { value: "1", isChecked: false },
        { value: "2", isChecked: true },
      ]);
    });

    test("Hidden values are not displayed", async () => {
      hideRows(model, [3]);
      await openFilterMenu();
      const values = getFilterMenuValues();
      expect(values.map((val) => val.value)).toEqual(["(Blanks)", "1"]);
    });

    test("Values hidden by another filter are not displayed", async () => {
      updateFilter(model, "B1", ["B2", "B3"]);
      await openFilterMenu();
      const values = getFilterMenuValues();
      expect(values.map((val) => val.value)).toEqual(["(Blanks)", "2"]);
    });

    test("Can hover mouse to select items", async () => {
      await openFilterMenu();
      const listItems = fixture.querySelectorAll(".o-filter-menu-value");
      expect(listItems[0].classList.contains("selected")).toBeFalsy();

      listItems[0].dispatchEvent(new Event("mousemove", { bubbles: true }));
      await nextTick();
      expect(listItems[0].classList.contains("selected")).toBeTruthy();
    });

    test("Clicking on values check and uncheck them", async () => {
      await openFilterMenu();
      expect(getFilterMenuValues()[0]).toEqual({ value: "(Blanks)", isChecked: true });
      await simulateClick(".o-filter-menu-value");
      expect(getFilterMenuValues()[0]).toEqual({ value: "(Blanks)", isChecked: false });
      await simulateClick(".o-filter-menu-value");
      expect(getFilterMenuValues()[0]).toEqual({ value: "(Blanks)", isChecked: true });
    });

    test("Confirm button updates the filter with formatted cell value", async () => {
      setFormat(model, "m/d/yyyy", target("A4"));
      expect(model.getters.getFilterValues({ sheetId, col: 0, row: 0 })).toEqual([]);
      await openFilterMenu();
      await simulateClick(".o-filter-menu-value:nth-of-type(2)");
      await simulateClick(".o-filter-menu-value:nth-of-type(3)");
      await simulateClick(".o-filter-menu-button-primary");
      expect(model.getters.getFilterValues({ sheetId, col: 0, row: 0 })).toEqual(["1", "1/1/1900"]);
    });

    test("Cancel button don't save the changes", async () => {
      expect(model.getters.getFilterValues({ sheetId, col: 0, row: 0 })).toEqual([]);
      await openFilterMenu();
      await simulateClick(".o-filter-menu-value:nth-of-type(1)");
      await simulateClick(".o-filter-menu-value:nth-of-type(2)");
      await simulateClick(".o-filter-menu-button-cancel");
      expect(model.getters.getFilterValues({ sheetId, col: 0, row: 0 })).toEqual([]);
    });

    test("Can clear and select all", async () => {
      await openFilterMenu();
      expect(getFilterMenuValues()).toEqual([
        { value: "(Blanks)", isChecked: true },
        { value: "1", isChecked: true },
        { value: "2", isChecked: true },
      ]);
      await simulateClick(".o-filter-menu-action-text:nth-of-type(2)");
      expect(getFilterMenuValues()).toEqual([
        { value: "(Blanks)", isChecked: false },
        { value: "1", isChecked: false },
        { value: "2", isChecked: false },
      ]);
      await simulateClick(".o-filter-menu-action-text:nth-of-type(1)");
      expect(getFilterMenuValues()).toEqual([
        { value: "(Blanks)", isChecked: true },
        { value: "1", isChecked: true },
        { value: "2", isChecked: true },
      ]);
    });

    test("Hitting esc key correctly closes the filter menu", async () => {
      await openFilterMenu();
      expect(fixture.querySelectorAll(".o-filter-menu")).toHaveLength(1);
      await keyDown("Escape");
      expect(fixture.querySelectorAll(".o-filter-menu")).toHaveLength(0);
    });

    describe("Search bar", () => {
      test("Can filter values with the search bar", async () => {
        await openFilterMenu();
        const searchInput = fixture.querySelector(".o-filter-menu input") as HTMLInputElement;
        searchInput.value = "1";
        searchInput.dispatchEvent(new Event("input", { bubbles: true }));
        await nextTick();
        expect(getFilterMenuValues().map((v) => v.value)).toEqual(["1"]);
      });

      test("Clicking on values focus the search input", async () => {
        await openFilterMenu();
        const searchInput = fixture.querySelector(".o-filter-menu input");
        expect(document.activeElement).not.toBe(searchInput);
        await simulateClick(".o-filter-menu-value");
        expect(document.activeElement).toBe(searchInput);
      });

      test("Search bar uses fuzzy search", async () => {
        setCellContent(model, "A2", "Florida");
        setCellContent(model, "A3", "Alaska");
        setCellContent(model, "A4", "Texas");
        setCellContent(model, "A5", "Illinois");

        await openFilterMenu();
        const searchInput = fixture.querySelector(".o-filter-menu input") as HTMLInputElement;
        searchInput.value = "lo";
        searchInput.dispatchEvent(new Event("input", { bubbles: true }));
        await nextTick();
        expect(getFilterMenuValues().map((v) => v.value)).toEqual(["Florida", "Illinois"]);
      });

      test("Can use up/down arrow keys to select items", async () => {
        await openFilterMenu();
        const searchInput = fixture.querySelector(".o-filter-menu input") as HTMLInputElement;

        searchInput.dispatchEvent(new KeyboardEvent("keydown", { key: "ArrowDown" }));
        await nextTick();
        const listItems = fixture.querySelectorAll(".o-filter-menu-value");
        expect(listItems[0].classList.contains("selected")).toBeTruthy();

        searchInput.dispatchEvent(new KeyboardEvent("keydown", { key: "ArrowDown" }));
        await nextTick();
        expect(listItems[1].classList.contains("selected")).toBeTruthy();

        searchInput.dispatchEvent(new KeyboardEvent("keydown", { key: "ArrowUp" }));
        await nextTick();
        expect(listItems[0].classList.contains("selected")).toBeTruthy();
      });

      test("Can press enter to select/unselect items", async () => {
        await openFilterMenu();
        const searchInput = fixture.querySelector(".o-filter-menu input") as HTMLInputElement;

        searchInput.dispatchEvent(new KeyboardEvent("keydown", { key: "ArrowDown" }));
        await nextTick();

        searchInput.dispatchEvent(new KeyboardEvent("keydown", { key: "Enter" }));
        await nextTick();
        expect(getFilterMenuValues()[0].isChecked).toBeFalsy();

        searchInput.dispatchEvent(new KeyboardEvent("keydown", { key: "Enter" }));
        await nextTick();
        expect(getFilterMenuValues()[0].isChecked).toBeTruthy();
      });
    });
  });

  test("Sort filter", async () => {
    createFilter(model, "A10:B15");
    setCellContent(model, "A10", "header");
    setCellContent(model, "A11", "olà");
    setCellContent(model, "A12", "1");
    setCellContent(model, "A13", "-1");
    setCellContent(model, "A14", "2");

    setCellContent(model, "B10", "header");
    setCellContent(model, "B11", "");
    setCellContent(model, "B12", "ab");
    setCellContent(model, "B13", "ba");
    setCellContent(model, "B14", "ca");
    await nextTick();

    await openFilterMenu();
    await simulateClick(".o-filter-menu-item:nth-of-type(1)");
    expect(getCellsObject(model, sheetId)).toMatchObject({
      A10: { content: "header" },
      A11: { content: "-1" },
      // A12 : empty cell, we give it a value of 0
      A13: { content: "1" },
      A14: { content: "2" },
      A15: { content: "olà" },
      B10: { content: "header" },
      B11: { content: "ba" },
      B13: { content: "ab" },
      B14: { content: "ca" },
    });

    await openFilterMenu();
    await simulateClick(".o-filter-menu-item:nth-of-type(2)");
    expect(getCellsObject(model, sheetId)).toMatchObject({
      A10: { content: "header" },
      A11: { content: "olà" },
      A12: { content: "2" },
      A13: { content: "1" },
      A15: { content: "-1" },
      B10: { content: "header" },
      B12: { content: "ca" },
      B13: { content: "ab" },
      B15: { content: "ba" },
    });
  });
});
