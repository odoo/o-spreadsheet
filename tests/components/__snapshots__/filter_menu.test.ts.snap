// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter menu component Filter Tests Filter menu is correctly rendered 1`] = `
<div
  class="o-filter-menu d-flex flex-column bg-white"
>
  <div>
    <div
      class="o-filter-menu-item"
    >
       Sort ascending (A ⟶ Z) 
    </div>
    <div
      class="o-filter-menu-item"
    >
       Sort descending (Z ⟶ A) 
    </div>
  </div>
  <div
    class="o-separator"
  />
  <div
    class="o-filter-menu-actions"
  >
    <div
      class="o-filter-menu-action-text"
    >
      Select all
    </div>
    <div
      class="o-filter-menu-action-text"
    >
      Clear
    </div>
  </div>
  <div
    class="position-relative"
  >
    <input
      class="w-100"
      placeholder="Search..."
      type="text"
    />
    <i
      class="o-search-icon position-absolute"
    >
      <svg
        class="search-icon"
        focusable="false"
        height="10"
        viewBox="0 0 512 512"
        width="10"
      >
        <path
          d="M500.3 443.7l-119.7-119.7c27.22-40.41 40.65-90.9 33.46-144.7C401.8 87.79 326.8 13.32 235.2 1.723C99.01-15.51-15.51 99.01 1.724 235.2c11.6 91.64 86.08 166.7 177.6 178.9c53.8 7.189 104.3-6.236 144.7-33.46l119.7 119.7c15.62 15.62 40.95 15.62 56.57 0C515.9 484.7 515.9 459.3 500.3 443.7zM79.1 208c0-70.58 57.42-128 128-128s128 57.42 128 128c0 70.58-57.42 128-128 128S79.1 278.6 79.1 208z"
          fill="#949494"
        />
      </svg>
    </i>
  </div>
  <div
    class="o-filter-menu-list d-flex flex-column rounded"
  >
    <div
      class="o-filter-menu-item o-filter-menu-value"
    >
      <div>
        <div
          class="o-filter-menu-value-checked"
        >
          <span>
            ✓
          </span>
        </div>
      </div>
      <div
        class="o-filter-menu-value-text text-truncate"
      >
        (Blanks)
        
        
      </div>
    </div>
    <div
      class="o-filter-menu-item o-filter-menu-value"
    >
      <div>
        <div
          class="o-filter-menu-value-checked"
        >
          <span>
            ✓
          </span>
        </div>
      </div>
      <div
        class="o-filter-menu-value-text text-truncate"
      >
        
        1
        
      </div>
    </div>
    <div
      class="o-filter-menu-item o-filter-menu-value"
    >
      <div>
        <div
          class="o-filter-menu-value-checked"
        >
          <span>
            ✓
          </span>
        </div>
      </div>
      <div
        class="o-filter-menu-value-text text-truncate"
      >
        
        2
        
      </div>
    </div>
    
    
    
  </div>
  <div
    class="o-filter-menu-buttons d-flex justify-content-end"
  >
    <div
      class="o-filter-menu-button o-filter-menu-button-cancel"
    >
       Cancel 
    </div>
    <div
      class="o-filter-menu-button o-filter-menu-button-primary"
    >
       Confirm 
    </div>
  </div>
</div>
`;
